src/scm.ts

```diff
@@ -0,0 +1,384 @@
+  maxResults?: number;
+}
+
+export interface FileContents {
+  sha: string;
+  content: string;
+  parsedContent: string;
+  mode: string;
+}
+interface FileDiff {
+  readonly mode: '100644' | '100755' | '040000' | '160000' | '120000';
+  readonly content: string | null;
+  readonly originalContent: string | null;
+}
+interface FileDiff {
+  readonly mode: '100644' | '100755' | '040000' | '160000' | '120000';
+  readonly content: string | null;
+  readonly originalContent: string | null;
+}
+export type ChangeSet = Map<string, FileDiff>;
+
+export interface ReleaseOptions {
+  draft?: boolean;
+  prerelease?: boolean;
+}
+export interface ScmRelease {
+  name?: string;
+  tagName: string;
+  sha: string;
+  notes?: string;
+  url: string;
+  draft?: boolean;
+  uploadUrl?: string;
+}
+
+export interface Scm {
+  readonly repository: Repository;
+  /**
+   * Returns the list of commits to the default branch after the provided filter
+   * query has been satified.
+   *
+   * @param {string} targetBranch Target branch of commit
+   * @param {CommitFilter} filter Callback function that returns whether a
+   *   commit/pull request matches certain criteria
+   * @param {CommitIteratorOptions} options Query options
+   * @param {number} options.maxResults Limit the number of results searched.
+   *   Defaults to unlimited.
+   * @param {boolean} options.backfillFiles If set, use the REST API for
+   *   fetching the list of touched files in this commit. Defaults to `false`.
+   * @returns {Commit[]} List of commits to current branch
+   * @throws {ScmAPIError} on an API error
+   */
+  commitsSince(
+    targetBranch: string,
+    filter: CommitFilter,
+    options?: CommitIteratorOptions
+  ): Promise<Commit[]>;
+
+  /**
+   * Iterate through commit history with a max number of results scanned.
+   *
+   * @param {string} targetBranch target branch of commit
+   * @param {CommitIteratorOptions} options Query options
+   * @param {number} options.maxResults Limit the number of results searched.
+   *   Defaults to unlimited.
+   * @param {boolean} options.backfillFiles If set, use the REST API for
+   *   fetching the list of touched files in this commit. Defaults to `false`.
+   * @yields {Commit}
+   * @throws {ScmAPIError} on an API error
+   */
+  mergeCommitIterator(
+    targetBranch: string,
+    options?: CommitIteratorOptions
+  ): AsyncGenerator<Commit, void, void>;
+
+  /**
+   * Get the list of file paths modified in a given commit.
+   *
+   * @param {string} sha The commit SHA
+   * @returns {string[]} File paths
+   * @throws {ScmAPIError} on an API error
+   */
+  getCommitFiles(sha: string): Promise<string[]>;
+
+  /**
+   * Iterate through merged pull requests with a max number of results scanned.
+   *
+   * @param {number} maxResults maxResults - Limit the number of results searched.
+   *   Defaults to unlimited.
+   * @yields {PullRequest}
+   * @throws {ScmAPIError} on an API error
+   */
+  pullRequestIterator(
```

Comment:
Whatever SCM we build would need to have a notion of pull requests (e.g. a change set in gerritt). Otherwise, release-please fundamentally wouldn't work -- it needs to create a release pull request of some kind.